python/src/python-module.cc

```diff
@@ -0,0 +1,75 @@
+#include <Python.h>
+
+#include "internal/eval.hh"
+#include "internal/ptr.hh"
+
+#include <eval.hh>
+#include <globals.hh>
+#include <shared.hh>
+
+namespace nix::python {
+
+#define _public_ __attribute__((visibility("default")))
+
+PyObject * ThrownNixError = nullptr;
+PyObject * NixError = nullptr;
+
+static PyMethodDef NixMethods[] = {
+    {"eval", (PyCFunction) eval, METH_VARARGS | METH_KEYWORDS, "Eval nix expression"}, {NULL, NULL, 0, NULL}};
```

Comment:
```suggestion
    {"evalDeep", (PyCFunction) eval, METH_VARARGS | METH_KEYWORDS, "Eval nix expression"}, {NULL, NULL, 0, NULL}};
```

I think this is better than `evalStrict`, but we should definitely reserve `eval`, because this is not a very general interface to the evaluator.

Reply:
I'm considering changing the API to a `nix.callExprString("{ a }: a", args=dict(a = 0))`, where `call` indicates that a function is called (compared to the current state which takes inputs from a static scope) and `String` indicates the function takes a string directly (instead of a file or an already-parsed expression)

Reply:
A `Value`-based API would be nicer.

 - `nix.valueFromExprString(s)`
 - `nix.call(fn, arg)`
 - `nix.autoCall(fn, attrs)`
 - `nix.toPythonStrict(value)`

`callExprString` could be implemented in Python in terms of those functions, but I don't think we even really need such a convenience function.

Crucially though, it's a more powerful interface this way, that enables use cases that aren't just a re-do of execve(nix-instantiate). The main benefits I see are that laziness can be exploited (even without a `toPythonLazy`), and more interesting data flows become possible in applications such as nixops, that are currently restricted to "evaluation phases", which are quite inconvenient and a bit messy and buggy.

Reply:
Agree with @roberth on this.

Reply:
> * nix.valueFromExprString(s)

Note that it's not just _parsing_ because that would go from string to `Expr`. Adding an `Expr` type would be useful for some advanced use cases, like doing some analyses, but it's excessive for now.

Reply:
@roberth you can actually return custom objects with a `__call__` implementation, so `nix.eval("x: x")(1)` _could_ work in the future.

Reply:
Should `__call__` force the result like `callFunction` or return an `App` thunk though? The latter is more general.

Reply:
> A `Value`-based API would be nicer.

Making such changes to the API would probably take me a lot longer to implement than just a simple adjustment. The current API is also nicely simple because it entirely avoids the need to persist C++ values in Python. We also don't currently have any concrete use cases for a more extensive API, for my use case the existing one is enough.

> I think for your motivating use case, a test framework, it's also nice not to have implicit auto-calling.

The proposed `nix.callExprString("{ a }: a", args=dict(a = 0))` shouldn't be implicit auto-calling: The passed string should have to be a callable expression, and `args` should be passed to this function as an argument directly. Which now makes me think it should be `arg` instead, because this should also be possible:

```python
nix.callExprString("a: a + 1", arg=0)
```

What do you think of that? This would be a fairly easy change I can make.

Reply:
Hmm, seems I was wrong to bring up auto-calling, but currently it's not even a function call. Is that something you want to introduce? Right now it's more like the [Python `eval`](https://docs.python.org/3/library/functions.html#eval).

> We also don't currently have any concrete use cases for a more extensive API

Not all considerations for a design have to be concrete use cases. We can pretend that the NixOps case doesn't exist, but that'd be foolish. The function does too many things to have a good name, because it is conceptually redundant. Its existence is an artefact of the implementation, and not a good design.
This kind of function leads to a forest of options and copies with variations that is as least as complex as a composable API, but without the flexibility and clarity. It should be removed as soon as possible.
In my view, almost all effort put into this function is a waste. Only the tests are useful because they can be adapted to a better interface.
Please consider implementing a sustainable, value-based interface.
There's an [official tutorial](https://docs.python.org/3/extending/newtypes_tutorial.html), if that helps. I'd like to help directly, but it will take a while before I have time for it.

Reply:
> Hmm, seems I was wrong to bring up auto-calling, but currently it's not even a function call. Is that something you want to introduce? Right now it's more like the [Python `eval`](https://docs.python.org/3/library/functions.html#eval).

Yes I want to do a slight API adjustment, because I find the idea of using the static scope to inject values very bad (that's what the current API does, similarly to `builtins.importScoped`), and also because that means the given strings aren't actually valid top-level Nix expressions (they have free variables). That would probably also make your proposed API harder to implement in the future. And all of that can be fixed just by changing the current API to something like `nix.callExprString("{ a }: a", arg=dict(a = 0))`, and it will take me like an hour at most (already much less than I spent discussing this now).

The Nix Team [agreed](https://github.com/NixOS/nix/pull/7735#issuecomment-1415853274) to make the bindings experimental and to merge it when ready, with the intent of limiting the bikeshedding and just getting a first approximation merged. What you're proposing would significantly delay this PR and be a deviation of that decision.

Note that there's still some important things to do for this PR even with the single-function API, most notably more testing and documentation, I'm gonna make this single-function API as future-proof as I can so that hopefully we can agree to just mark it stable soon.

> The function does too many things to have a good name, because it is conceptually redundant. Its existence is an artefact of the implementation, and not a good design.
> This kind of function leads to a forest of options and copies with variations that is as least as complex as a composable API, but without the flexibility and clarity. It should be removed as soon as possible.

Hard disagree:
- Composability isn't in conflict with convenient functions combining functionality. Removing the ability to do `nix.callExprString("{ a }: a", arg=dict(a=0))` in favor of the proposed API, so users have to write `nix.toPythonStrict(nix.call(nix.valueFromExprString("{ a }: a"), arg= nix.fromPython(dict(a=0))))` instead is just pointless.
- No it doesn't have to lead to a forest of variations, because as soon as somebody wants to add `nix.callExprFile` you can complain and tell them that we should have `nix.parseFile`, `nix.parseString` and `nix.callExpr` instead and define `nix.callExprString` as `nix.callExpr(nix.parseString())` for backwards compatibility. If you want I can add a note on this in the hacking section, for people wanting to extend the API in the future.

Reply:
I think we disagree about whether this library is about bindings or providing some sort of high-level api. The latter seems like a fool's errand to me.


> * Composability isn't in conflict with convenient functions combining functionality.

They do impose a maintenance cost and users have to weed through the docs to find the accurate bits. I don't want the bindings to end up looking like [Haskell `System.Process`](https://hackage.haskell.org/package/process-1.6.16.0/docs/System-Process.html).

> * Removing the ability to do `nix.callExprString("{ a }: a", arg=dict(a=0))` in favor of the proposed API, so users have to write `nix.toPythonStrict(nix.call(nix.valueFromExprString("{ a }: a"), arg= nix.fromPython(dict(a=0))))` instead is just pointless.

Whose convenience? The latter is easier to understand. You can tell exactly what's going on. Nobody needs to learn about the non-obvious `callExprString` semantics. If the user's own code gets repetitive, they can define their own abstraction that's most likely better than `callExprString`.
If we follow @yorickvP's suggestion, the code is even terse enough to repeat without introducing such an indirection:

```python
nix.evalString("a: a + 1")(0).forceDeep()
```

as opposed to

```python
nix.callExprString("a: a + 1", arg=0)
```

The syntax is obviously not as explicit as what I suggested, but it seems that these are standard methods that python objects are expected to support, so it seems that would be explicit enough in the mind of a python programmer. More important is that the interface composes, instead of exposing arbitrary combinations.

> bindings experimental

I agree that something is better than nothing, but if the only function is one we know we don't really need, the net value of merging this is barely positive, but probably positive nonetheless.

Reply:
> I think we disagree about whether this library is about bindings or providing some sort of high-level api

Bindings don't have to be low-level, I consider this a high-level binding, where the level is the same as the CLI.

In any case, I appreciate the thoughts and ideas, even if I don't fully agree with them, but I don't think it makes sense to discuss here if it's not a blocker for this PR.

Reply:
I am still with @roberth here

> where the level is the same as the CLI.

CLIs specifically make the tradeoff of doing "evil" things for the sake of *live typing* ergonomics that would not be acceptable in a programmable interface. Unless I am mistaken, these bindings are not meant mainly for ad-hoc interactive usage in the Python REPL, but committed code as part of larger python programs. As such, regular notions of compositional, "do one thing and do it well", etc. are important and "terseness at all costs" is not.

Reply:
So yeah I can see what you mean, and I agree, such a composable interface would be nice. I'm in full support for such an interface in a future version of the API. However I do not think this single function here is as bad as you make it out to be, it's not like it will ruin the Python bindings forever, especially since it won't even be stable.

Reply:
Depending on how much time @infinisil has available for this, I'd be ok merging and releasing this, so we can validate the broader aspects of python bindings without getting bogged down and potentially blocking his use case for the bindings.
We should do some expectation management because it's experimental anyway.